### DISCUSSION TITLE
`sizeof` için `%d` uyarısı

### DISCUSSION THREAD

**Discussion by berilkones** [19/classroom#69] [C] [answered]
Hocam sizeof komutu ile ilgili çalışma yaparken ilgili satırda %d komutu kullanınca yazmış olduğum kod hata verdi. %d kullanmak yerine %ld yazdığımda ise kodum hatasız bir şekilde çalıştı. Konuyla ilgili araştırma yaptığımda kodun çalıştığım platform dışında başka yerlerde de %d ile çalışabildiğini gördüm. Çalışmamasının sebebi nedir?

![Ekran görüntüsü 2024-03-24 202000](https://github.com/19/classroom/assets/151151773/4755fae2-3cdb-47e7-bd55-8b686283d0b8)
![Ekran görüntüsü 2024-03-24 203618](https://github.com/19/classroom/assets/151151773/8a5665cc-24bf-4e43-a061-655418a197d3)

**Comment by roktas** [comment, marked as answer]
Gönderdiğin ilk ekran görüntüsünü dikkatli incele. Aldığın bir "hata" (error) değil bir "uyarı" (warning). Yani derleme uyarıya rağmen başarılı (program üretiliyor). Sana söylemeye çalıştığı şey `sizeof(beril)`'in normal bir tamsayı değil uzun biçimde işaretsiz bir tamsayı ürettiği. `4` değeri iki biçimde de geçerli olduğundan sonuç değişmiyor ama yine de derleyici uyarmayı ihmal etmiyor.

`sizeof` neden (işaretli) olağan bir tamsayı değil de daha uzun bir tamsayı dönüyor? Boyutunu öğrenmeye çalıştığın nesne çok büyük boyutlu olabilir de ondan, öyle ki normal bir tamsayı bu değerde taşma gösterebilir (maksimum tamsayıdan da büyük bir boyut değeri olabilir). Neden işaretsiz? "İşaret için ayrılan biti de kullanalım ki temsil edilebilecek boyut değeri daha da büyüsün (2 kat daha büyük olsun)" diye düşünülmüş.

Not: Aldığınız hata için ekran görüntüsü kullanabilirsiniz buna itirazım yok. Fakat kaynak kodu ekran görüntüsü yerine lütfen benim aşağıda yaptığım gibi yazın. Aksi halde kodu görüntüye bakarak tekrar yazmak gerekiyor. 

```c
#include <stdio.h>

int main(void)
{
	int integer;

	printf("Değişken bellekte %d yer kaplar\n", sizeof(integer));

	return 0;
}
```
